[2022.10.22] Feat: SideBar 기능 추가

Y-SeonJong opened the pull request:
### task card 제목
Fix: SideBar 기능 추가
### Task

- task card 링크
https://www.notion.so/vanillacoding/Diffing-50571a196d5b40d9803c2fd870a0f7c1
### Description
![image](https://user-images.githubusercontent.com/76609548/196759794-c2d93a5b-6512-47f5-826b-e242e45f8f66.png)

- 작업 내용
SIdeBar에 조건에 맞게 슬라이드 변경 사항 배치
슬라이드 전체 추가 혹은 삭제 되었을 경우 각 칸에 맞게 
배치후 
Slide Item이 존재하지 않기 때문에 Slide ID 에 CheckBox 부여

한 슬라이드 내에서 
수정, 추가, 삭제 의 경우
수정의 경우 : ADD, UNDO의 칸 모두에 부여
추가의 경우 : ADD 칸에만 설정
삭제의 경우 : UNDO 칸에 설정

// 작업 내용 추가
Diffing 된 결과 값의 Type들을 config 폴더에 따로 명시,
util에 원하는 타입에 맞게 데이터를 filter 할 수 있는 기능 구현

최상단에서 필터링 되도록 구현
### Point

- 리뷰어가 집중해줬으면 하는 부분(컨펌 필요한 부분)
변수명....!!!! 이 너무 어색하네요..

### ETC
아직 CheckBox에 대한 세세한 설정은 작업 더 진행 예정
- 그외 추가사항이 있다면

Y-SeonJong committed:
Feat: 사이드바 기능 추가
사이드바를 ADD, UNDO (이름은 추후 수정)로 나누어 구분 했습니다.
사이드 바 내의 목록의 경우 크게 Slide 전체와 Slide 세부 사항이
존재합니다.
슬라이드 전체 추가, 삭제의 경우 Slide 자체에 CheckBox를
나머지 경우 Slide 하위 태그에 넣은 상태입니다.
ADD 에는 텍스트 내용이 추가, modified
undo에는 삭제된 텍스트 수정이 들어갑니다.

Y-SeonJong committed:
Style: Commit 중 console.log 보여서 삭제 후 Push

Donghyeon0908 commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+import mockDiffData from "../../mock"; // Todo: Mock Data에서 삭제, 현재 로직은 Mock Data로직 변경 필요.
+
 function SideBar() {
+  const [mockDatas, setMockDatas] = useState({});
+  const [diffDatas, setDiffDatas] = useState([]);
+  const [addedDatas, setAddedDatas] = useState([]);
+  const [modifiedDatas, setModifiedDatas] = useState([]);
+  const [deleteDatas, setDeleteDatas] = useState([]);
```
data 자체가 복수형이라 datas는 어색해보여요

Y-SeonJong commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
 import SideBarSlideList from "./SideBarSlideList";
 
+import mockDiffData from "../../mock"; // Todo: Mock Data에서 삭제, 현재 로직은 Mock Data로직 변경 필요.
+
```
개행한줄을 더 했네요 수정했습니다.

Y-SeonJong commented on src/components/SideBar/SideBarItems.js:
```diff
@@ -0,0 +1,61 @@
+
+function SideBarItems({ categorizedData, type }) {
+  const [modifiedTitles, setModifiedTitles] = useState([]);
+  const [finialTitles, setFinialTitles] = useState([]);
```
이 부분에 대해서 변수명을 
수정 데이터 중에서도 분류가 완료 된 수정 데이터인데 
Title은 일단 이상하네요..
근데 마땅한 변수명이 생각이 나지 않아요 ...

Donghyeon0908 commented on src/mock.js:
```diff
@@ -0,0 +1,356 @@
+        isChecked: false,
+      },
+      "TextBox 7": {
+        diff: "deleted",
+        isChecked: false,
+      },
+      "그림 8": {
+        diff: "modified",
+        isChecked: false,
+      },
+    },
+  },
+  3193319780: {
+    diff: "modified",
+    items: {
+      "TextBox 2": {
+        diff: "modified",
+        isChecked: false,
+      },
+      "TextBox 4": {
+        diff: "added",
+        isChecked: false,
+      },
+      "TextBox 6": {
+        diff: "modified",
+        isChecked: false,
+      },
+      "TextBox 7": {
+        diff: "deleted",
+        isChecked: false,
+      },
+      "그림 8": {
+        diff: "modified",
+        isChecked: false,
+      },
+    },
+  },
+};
+
+export default mockDiffData;
```
지금은 로직을 확인하는데 mock데이터가 있어서 이해하는데 도움을 주지만 merge하시기전에 지우고 새로 commit 하시는건 어떨까요? dev브런치에 있을 이유는 없어보여요

Y-SeonJong replied:
넵~!!

Donghyeon0908 commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+  const [modifiedDatas, setModifiedDatas] = useState([]);
+  const [deleteDatas, setDeleteDatas] = useState([]);
+
+  useEffect(() => {
+    if (!mockDiffData) return;
+    setMockDatas(mockDiffData); // Todo Data 받는 양식 mockdata가 아닌 merge 데이터 받아오는 작업
+  }, []);
```
주석에 merge 데이터를 받아온다는게 잘 이해가 안되서 설명부탁드려요~

Y-SeonJong replied:
아 설명을 제가 잘못 쓴 것 같아요!
Mock Data와 같은 /api/compare 에서 post 요청으로 compare 데이터를 받아온다가 맞는 설명 같아요!

Donghyeon0908 commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+    setMockDatas(mockDiffData); // Todo Data 받는 양식 mockdata가 아닌 merge 데이터 받아오는 작업
+  }, []);
+
+  useEffect(() => {
+    setDiffDatas(Object.keys(mockDatas));
+  }, [mockDatas]);
+
+  useEffect(() => {
+    setAddedDatas(diffDatas.filter((key) => mockDatas[key].diff === "added"));
+    setModifiedDatas(
+      diffDatas.filter((key) => mockDatas[key].diff === "modified"),
+    );
+    setDeleteDatas(
+      diffDatas.filter((key) => mockDatas[key].diff === "deleted"),
+    );
+  }, [mockDatas, diffDatas]);
```
`useEffect `를 사용하신 이유가 있으실까요? 궁금합니다!

Y-SeonJong replied:
아 기능을 적다 보니까... 나중에 되면 파일을 교체 했을 때 diffData가 변할 수 있으니까라고 생각해서 useEffect()를 적었는데 생각해보니까

파일을 교체하는 경우가 없었네요..

Donghyeon0908 commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
-        <SideBarHeader>Create</SideBarHeader>
-        <SideBarSlideList />
+        <SideBarHeader>ADD</SideBarHeader>
+        {addedDatas.map((key) => (
+          <SideBarSlideList
+            categorizedData={mockDatas[key]}
+            title={key}
+            key={key}
```
전체적으로 key라는 네이밍이 조금 어색해보이는 어떻게 생각하시나요?

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+
 function SideBar() {
+  const [mockDatas, setMockDatas] = useState({});
+  const [diffDatas, setDiffDatas] = useState([]);
```
state로 둘 이유가 없어 보이는 값들인 것 같아요.. 각각의 값들은 mock data에서 연산이 되어서 나오는 값들이지, 따로 state로 둘 이유가 없지 않을까요?

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+  }, [mockDatas]);
+
+  useEffect(() => {
+    setAddedDatas(diffDatas.filter((key) => mockDatas[key].diff === "added"));
```
filter method의 첫번째 파라미터로 key라고 작성하신 이유가 있을까요? filter 메소드의 첫번째 값은 현재의 element 값인데, diffData 정도로 해도 충분할 것 같아요

Y-SeonJong replied:
생각해보니까 diffData 자체를 diffSlideIdList로 하는게 더 좋겠네요
DiffData는 데이터의 전체 내용을 가져온게 diffData인데 
현재 제가 작성한 로직에서는 diffData엔 슬라이드 이름의 목록들만 나열해서요!

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+        <SideBarHeader>ADD</SideBarHeader>
+        {addedDatas.map((key) => (
+          <SideBarSlideList
+            categorizedData={mockDatas[key]}
```
변수명이 component 관점에서 지어진다면 더 좋을 것 같아요. component 입장에서는 categorizedData가 아니라 slideData 정도로 해도 좋지 않을까요?

Y-SeonJong replied:
분류가 되었다는 의미를 명시해 주고 싶은데 SlideData로 만은 분류가 되었다는걸 나타냈다고 보기에는 어려워서 더 좋은 변수명은 없을까요?

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+            categorizedData={mockDatas[key]}
+            title={key}
+            key={key}
+            type="added"
```
type과 같은 부분은 상수화를 시켜주도록 해도 좋을 것 같아요~

0jaaack commented on src/components/SideBar/SideBarItems.js:
```diff
@@ -0,0 +1,61 @@
+  const [finialTitles, setFinialTitles] = useState([]);
+
+  useEffect(() => {
+    setModifiedTitles(Object.keys(categorizedData.items));
```
단순히 prop으로 가져온 값을, state로 관리할 필요가 없을 것 같아요~

0jaaack commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,22 +1,25 @@
-        <SideBarItem>
+      {!categorizedData.items ? (
+        <SideBarLittleHeader>
           <InputLabel>
```
InputLabel이라는 component명보다는 ListItemLabel, ItemLabel 정도로 짓는 게 목적성이 뚜렷한 것 같은데, 선종님은 어떻게 생각하시나요?

0jaaack commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,22 +1,25 @@
-      <SideBarItemList>
-        <SideBarItem>
+      {!categorizedData.items ? (
+        <SideBarLittleHeader>
```
LittleHeader 말고 MainHeader가 따로 있다면 쓸 수 있겠지만, 없기 때문에 조금 모호한 컴포넌트 명이 된 것 같아요. `SideBarListHeader` 정도로 해도 좋지 않을까요?

0jaaack commented on src/components/SideBar/SideBarItems.js:
```diff
@@ -0,0 +1,61 @@
+  }, [categorizedData, modifiedTitles, type]);
+
+  return (
+    <SideBarItem>
```
SideBarItems 컴포넌트의 Container로 SideBarItem을 두신 이유가 있을까요? 컨벤션에 맞지 않는 컴포넌트명인것 같습니다.

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -1,17 +1,73 @@
+  }, [mockDatas]);
+
+  useEffect(() => {
+    setAddedDatas(diffDatas.filter((key) => mockDatas[key].diff === "added"));
```
diff의 조건문을 `diff === "added" || diff === "modified"`로 해도 값을 구할 수 있지 않나요? 따로 modifiedDatas를 구할 이유는 없어보입니다~

Y-SeonJong replied:
어제 목데이터 가지고 작업을 할때 
두번째 목데이터가 나오기던에 첫 목데이터에서 added 안에 데이터 아이템이 있다고 가정을 해서 아이템 여부로 나눴습니다.

그래서 아이템 존재 여부로 전테, 일부 수정으로 했었네요

그리고 지금 드는 생각이 diff 타입을
added deleted modified(added, modified, deleted) 보다는

Wholeadded WholeDeleted modified(added ,modified ,deleted) 로 하는게 어떨까요

added 가 추가라는 개념에서는 동일하지만 전체추가와 단일추가는 완전히 다른 기능이라고 생각을 해서요!

0jaaack commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,22 +1,25 @@
-function SideBarSlideList({ sideBarData }) {
+function SideBarSlideList({ categorizedData, title, type }) {
   return (
     <SideBarDataContainer>
```
SideBarSlideListContainer로 component 명이 되어야 할 것 같아요~

0jaaack commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,22 +1,25 @@
-      <SideBarLittleHeader>슬라이드 제목 1</SideBarLittleHeader>
-      <SideBarItemList>
-        <SideBarItem>
+      {!categorizedData.items ? (
```
items라는 값이 있으면 item 슬라이드 데이터를 보여준다... 보다는 저희가 수정된 슬라이드에 한해서 슬라이드 내부 요소를 비교하기로 하였으니, diff의 값을 이용해 분류하는 게 좋아보여요~

0jaaack commented on src/components/SideBar/SideBarItems.js:
```diff
@@ -0,0 +1,61 @@
+  );
+}
+
+const SideBarItem = styled.li`
```
내부에 자식 태그들이 나열되는 태그인데, li 태그보다는 ul와 같은 태그를 사용하심이 좋아보입니다!

Y-SeonJong replied:
이게 어떤 소리인지 이해가 가지않네요..?

0jaaack reviewed (commented):
선종님~ 전체적으로 component 구조와 데이터 관리가 조금 더 깔끔하게 이루어진 상태로 컴포넌트들이 짜여지면 더욱 좋을 것 같아요...! 논의할 내용 있으시면 댓글 부탁드립니다~

Y-SeonJong committed:
Fix: 피드백 구조 반영
1. 상수화 관련 수정
2. map 관련 items => diff.type
3. gitignore에 mock.js 추가 하여 mock데이터 안올라가게끔 수정
4. 컨벤션 맞지 않는 부분 수정

Y-SeonJong committed:
Fix: mock 데이터 삭제
gitignore 에 mock.js 추가해서 올릴려 햇으나 삭제되지 않아
mock.js 삭제 후 commit

Donghyeon0908 commented on src/components/SideBar/SideBar.js:
```diff
@@ -3,15 +3,43 @@ import styled from "styled-components";
+  const diffSlideIdList = Object.keys(mockDiffData);
+  const addedData = diffSlideIdList.filter(
+    (slideId) =>
+      mockDiffData[slideId].diff === "added" ||
+      mockDiffData[slideId].diff === "modified",
+  );
+  const deletedData = diffSlideIdList.filter(
+    (slideId) =>
+      mockDiffData[slideId].diff === "deleted" ||
+      mockDiffData[slideId].diff === "modified",
+  );
+  const typeAdded = "added";
+  const typeDeleted = "deleted";
```
이곳에서 선언해주는것도 좋지만 src/config/constants에 따로 diffType을 만들어서 관리하는건 어떨까요? 그리고 지금 로직에선 타입 선언을 fliter보다 위에 두어`      mockDiffData[slideId].diff === typeDeleted ,` 이렇게 해주는게 더 좋아보이는데 다른분들 의견이 궁금합니다

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -3,15 +3,43 @@ import styled from "styled-components";
 import SideBarSlideList from "./SideBarSlideList";
 
 function SideBar() {
+  const diffSlideIdList = Object.keys(mockDiffData);
```
저희 현재는 mock data으로 구현하고 있지만, 기능 구현의 사항이므로 `diffData`라고 기재해주시면 감사하겠습니다~

0jaaack commented on src/components/SideBar/SideBarItemList.js:
```diff
@@ -0,0 +1,51 @@
+        modifiedData.items[title].diff === "added" ||
+        modifiedData.items[title].diff === "modified",
+    );
+  } else {
```
type이 꼭 "added"가 아니라면, "deleted"인 경우는 아닐 수도 있으므로, 케이스가 구체적으로 설정된 조건 문을 작성해주시면 좋을 것 같아요!

0jaaack commented on src/components/SideBar/SideBarItemList.js:
```diff
@@ -0,0 +1,51 @@
+  return (
+    <SideBarItemListContainer>
+      {finialTitles.map((title) => (
+        <SideBarItemListLabel key={title}>
```
ul 태그의 자식 태그로 li 태그가 있어야 하지 않을까요?? 🧐

0jaaack commented on src/components/SideBar/SideBar.js:
```diff
@@ -3,15 +3,43 @@ import styled from "styled-components";
 
 function SideBar() {
+  const diffSlideIdList = Object.keys(mockDiffData);
+  const addedData = diffSlideIdList.filter(
```
addedData로 단순히 id값들을 배열로 가진 다음, 하위에서 그  id 값으로 로직을 한 번 더 작성하고 있는데요. 하위에서 렌더링 될 값들을 상위에서 미리 다 계산해놓으신다면 굳이 복잡하게 로직을 안짜도 된다고 되지 않을까요?

상위에서
```
const addedSlides = [
  { slideId: "123" },
  { slideId: "456", items: [ "tx1", "tx2", "img1" ] },
];

```
이런 식의 값이 있다면, 하위에서 item 값을 가지기 위해 로직을 작성할 수고를 덜 수 있을 것 같아요 한 번 확인 부탁드리겠습니다~

0jaaack reviewed (commented):
선종님 로직 작성하신 부분 및 수정 반영된 부분 잘 보았습니다~ 아래 추가 리뷰사항도 한 번 읽어봐주시면 감사하겠습니다!

Y-SeonJong committed:
Fix: SideBar 구조 수정
최상위 컴포넌트에서 최대한 분류

Y-SeonJong commented on src/components/SideBar/SideBar.js:
```diff
@@ -2,16 +2,34 @@ import React from "react";
+
 function SideBar() {
+  const addedSlideData = filterDiffDataByType(diffData, DIFFING_TYPES.ADDITION);
+  const deletedSlideData = filterDiffDataByType(diffData, DIFFING_TYPES.DELE);
```
Type 이름 자체를
추가, 삭제, 수정으로 할려고 생각을 해서
각각의 명사형을 찾아서 넣었는데
DELE 자체가 삭제를 의미하더라고요. 혹시 이상한가요?

Donghyeon0908 replied:
DELE가 조금 어색한거같아요~ DELETED_TYPE은 이상한가요?

Y-SeonJong commented on src/components/SideBar/SideBarItemList.js:
```diff
@@ -0,0 +1,40 @@
+import styled from "styled-components";
+
+function SideBarItemList({ slideData }) {
+  const slideContents = Object.keys(Object.values(slideData)[0].items);
```
FinialTitle -> slideContents로
변경 이유 : 슬라이드 페이지 내에 있는 내용들을 표시해준다고 생각해서
slideContents로 설정

Y-SeonJong commented on src/components/SideBar/SideBarItemList.js:
```diff
@@ -0,0 +1,40 @@
+  list-style: none;
+`;
+
+const SlideBarItemList = styled.li``;
```
이렇게 li 태그를 추가해달라는게 맞을까요?

Donghyeon0908 replied:
현재 로직에선 `SideBarItemListContainer ` 컴포넌트가 `styled.ul` 이되어야할거같아요

Y-SeonJong commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,41 +1,40 @@
 
-function SideBarSlideList({ sideBarData }) {
+function SideBarSlideList({ slideData, type }) {
+  const slidePage = Object.keys(slideData);
```
SlideId 보다는 SlidePage로 하는건 어떻게 생각하시나요?
사이드바에 보여지는게 결과적으로 어떤 슬라이드 페이지에서 변경된건지라 생각해서 SlidePage로 변수를 선언했어요

Y-SeonJong reviewed (commented):
리뷰 부탁드리겠습니다~!!!!!!!!

Y-SeonJong committed:
Feat: 사이드바 리덕스 스토어 추가
리덕스 스토어를 추가 한 상태 리덕스 변환 도중 커밋 한번 추가
현재 비교 데이터들이 분류되어 add undo위치에 나타남

Y-SeonJong committed:
Fix: 목데이터 삭제
목데이터 삭제 및 개행 작업 확인

Y-SeonJong reviewed (commented):
리덕스로 상태 관리를 하려다 보니까 좀 막막해서 같이 생각해보면 좋을 것 같아서 먼저 커밋 한번 보냈습니다.!

Y-SeonJong reviewed (commented):
현재 상황은
체크 박스 클릭시 해당하는 값들의 isChecked 들이 개별적으로 false, true 로 변경 되는 상태이며
,
type : added > 상단
type: deleted> 하단
type: modified > modified 상단, 하단
type: modified > added 상단
type: modified > deleted 하단 으로 분류 해 논 상태입니다.

0jaaack commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,41 +1,53 @@
+          </SideBarSlideListLabel>
+        </SideBarListHeader>
+      ) : (
+        // eslint-disable-next-line react/jsx-no-useless-fragment
```
eslint 관련 주석은 굳이 안넣으셔도 좋을 것 같습니다!

Y-SeonJong replied:
아 이 주석을 넣은 이유가 주석을 안넣으면 에스렌트 오류 떠서 무시하려고 넣었어요

0jaaack commented on src/components/SideBar/SideBarSlideList.js:
```diff
@@ -1,41 +1,53 @@
+      ) : (
+        // eslint-disable-next-line react/jsx-no-useless-fragment
+        <>
+          {Object.keys(Object.values(slideData)[0].items).length !== 0 ? (
```
삼항 연산자를 null 조건으로 주기 보다는, AND 연산자(&&)를 활용하심이 어떨까요?

Donghyeon0908 reviewed (approved):
선종님 새벽까지 정말 고생많으셨습니다! 다른부분은 리뷰를 하면서 이야기하도록 해요

0jaaack commented on src/components/Main.js:
```diff
@@ -32,11 +51,12 @@ function Main() {
         </MainContainer>
       );
     case SEQUENCES.COMPARISION:
+      getDiffingResult();
```
diffing result를 여기서 fetch할 필요는 없을 것 같아요! 비교하기 버튼을 클릭했을 때 나와야 하니 이 부분은 옮겨지면 좋을 것 같습니다!

0jaaack reviewed (approved):
선종님 사이드바 관련해서 고생많으셨습니다! 시각화 관련해서 고생이 많으셨는데 그만큼 저희가 성장하고 좋은 시간이 되었다고 생각해요!!

Donghyeon0908 commented on src/components/Main.js:
```diff
@@ -2,19 +2,38 @@ import React from "react";
+    ({ pptData }) => pptData[VIEW_TYPES.COMPARABLE_FILE].dataId,
+  );
+
+  const getDiffingResult = async () => {
+    const diffingResult = await axios.post("/api/ppts/compare", {
+      originalPptId,
+      comparablePptId,
+    });
+
+    dispatch(
+      originalDiffData({
+        type: DIFFING_TYPES.DIFF_DATA,
+        data: diffingResult.data,
+      }),
```
비교하기 버튼을 누르면 서버로 비교 요청이 들어가니 `getDiffingResult `가 main에 있는건 어색해보이네요

Donghyeon0908 commented on src/components/MainSection/FileAttachment.js:
```diff
@@ -13,9 +13,11 @@ function FileAttachment({ fileType }) {
   const handleFileChanged = async (event) => {
     const pptData = await pptxParser(event.target.files[0]);
-    const response = await axios.post("/api/ppt/save", { pptData });
+    const response = await axios.post("/api/ppts/save", { pptData });
```
uri 수정해주셨군요 👍

Donghyeon0908 commented on src/components/SideBar/SideBarItemList.js:
```diff
@@ -0,0 +1,52 @@
+  );
+}
+
+const SideBarItemListContainer = styled.li`
```
여기는 `ul`로 되어야할거같은데 반영이 안되었네요
